setup.py

```diff
@@ -1,127 +1,209 @@
+    """
+    Read version information from the version file, if it exists.
+
+    Returns
+    -------
+    version : unicode
+        The full version, including any development suffix.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+
+    Raises
+    ------
+    EnvironmentError
+        If the version file does not exist.
+    """
+    version_info = runpy.run_path(VERSION_FILE)
+    return (version_info["version"], version_info["git_revision"])
 
-    with open(filename, "wt") as fp:
-        fp.write(template.format(version=VERSION,
-                                 full_version=fullversion,
-                                 git_revision=git_rev,
-                                 is_released=IS_RELEASED))
+
+def git_version():
+    """
+    Construct version information from local variables and Git.
+
+    Returns
+    -------
+    version : unicode
+        Package version.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+
+    Raises
+    ------
+    EnvironmentError
+        If Git is not available.
+    subprocess.CalledProcessError
+        If Git is available, but the version command fails (most likely
+        because there's no Git repository here).
+    """
+    git_count, git_revision = _git_info()
+    version_template = RELEASED_VERSION if IS_RELEASED else UNRELEASED_VERSION
+    version = version_template.format(
+        major=MAJOR, minor=MINOR, micro=MICRO, dev=git_count
+    )
+    return version, git_revision
+
+
+def resolve_version():
+    """
+    Process version information and write a version file if necessary.
+
+    Returns the current version information.
+
+    Returns
+    -------
+    version : unicode
+        Package version.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+    """
+    if os.path.isdir(GIT_DIRECTORY):
+        # This is a local clone; compute version information and write
+        # it to the version file, overwriting any existing information.
+        version = git_version()
+        print(u"Computed package version: {}".format(version))
+        print(u"Writing version to version file {}.".format(VERSION_FILE))
+        write_version_file(*version)
+    elif os.path.isfile(VERSION_FILE):
+        # This is a source distribution. Read the version information.
+        print(u"Reading version file {}".format(VERSION_FILE))
+        version = read_version_file()
+        print(u"Package version from version file: {}".format(version))
+    else:
+        raise RuntimeError(
+            u"Unable to determine package version. No local Git clone "
+            u"detected, and no version file found at {}.".format(VERSION_FILE)
+        )
+
+    return version
 
 
 if __name__ == "__main__":
-    write_version_py()
-    from envisage import __requires__, __version__
+    try:
+        version, git_revision = resolve_version()
+    except RuntimeError as err:
```

Comment:
This is the only difference between the changes introduced in this PR vs the changes in traits. This is necessary because if on a user machine

a) `envisage` does not have git tree and 
b) does not have an existing `envisage.version` file,

the package installation will fail, complaining about how the package version could not be determined.

IMO, it is better to complain and allow the package to be installed instead of not allowing installation.

Reply:
If this approach is preferable, we will need to fix this issue in traits.

Reply:
You may not be surprised to hear that I prefer the Traits approach. :-) If there's no git tree and no version file, then something's seriously wrong - essentially in that case, the environment is broken in some way, and that brokenness deserves investigation and fixing- carrying on won't fix the underlying issue, whatever it may be, and that underlying issue is likely to bite elsewhere.

I'd rather have reports of breakage in this case, so that if there's some peculiar way of installing this package then we find out about it and either cater for it explicitly if needed, or have a conversation about why it's the wrong way to do things.